### ISSUE TITLE
[TCP Server] Add io_uring support

### ISSUE THREAD

**Issue by ammarfaizi2** [TeaInside/teavpn2#23]
I am going to replace epoll with io_uring, however io_uring syscall is only available for Linux 5.1+, hence I will not drop the epoll support, I will just make a dispatcher to detect kernel version at runtime.

IOW, we can say like this:
```
if (linux_version >= 5.1) {
   // use io_uring
} else {
   // fallback to epoll
}
```

**Comment by ammarfaizi2** [comment]
I am going to delay this, and keep using epoll as primary events waiter, it seems io_uring isn't a stable syscall yet. There are several cases from liburing that cause the kernel hang and crashes.

The README file on liburing says like this:
_The bulk of liburing is actually regression/unit tests for both liburing and the kernel io_uring support. Please note that this suite isn't expected to pass on older kernels, and may even crash or hang older kernels!_

We will implement io_uring when Linux 5.13 has been released, maybe.

**Comment by ammarfaizi2** [comment]
Linux 5.12 has been released, I will prepare the io_uring support and hopefully it will be fully implemented when 5.13 is released.

**Comment by louvian** [comment]
Are you going to compile liburing statically?

**Comment by ammarfaizi2** [comment]
@louvian yes, io_uring is a collection of new syscalls. I can't depend on libc to use it as there is no interface for it, and probably the io_uring support won't be there in the near future. So I will compile and link the liburing statically to TeaVPN2 binary.

I don't think it is heavy to be statically linked. So that's fine.

And anyway, I think there is no distributed liburing shared object from Linux distros' official repo at the moment.

**Comment by ammarfaizi2** [comment]
Oh, it turned out there exists on my Ubuntu 21.04
```
ammarfaizi2@integral:~$ apt list 2>&1 | grep uring
liburing-dev/hirsute 0.7-3ubuntu1 amd64
liburing1/hirsute,now 0.7-3ubuntu1 amd64 [installed,automatic]
turing/hirsute 0.11~beta-3 all
ammarfaizi2@integral:~$ 
```
Well, I will reconsider the compilation strategy later.

**Comment by ammarfaizi2** [comment]
This will be on my plate for hacktoberfest.

**Comment by louvian** [comment]
> This will be on my plate for hacktoberfest.

Any update?

**Comment by ammarfaizi2** [comment]
> > This will be on my plate for hacktoberfest.
> 
> Any update?

Unfortunately, I did not have a good time for this. I did not even finish my hacktoberfest task.

Apparently, I'd love for that not to be the case, but there's just not enough time to get it done. At least my epoll implementation is stable enough, so I don't see that as a big problem.

One day we will have a full set of TeaVPN2 that contains poll(), epoll and io_uring. But it requires that someone writes them. So far, not much has happened there.
